core/modules/replica/ServiceManagementRequest.cc

```diff
@@ -0,0 +1,267 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/ServiceManagementRequest.h"
+
+// System headers
+#include <stdexcept>
+
+// Third party headers
+#include <boost/bind.hpp>
+#include <boost/date_time/posix_time/posix_time.hpp>
+
+// Qserv headers
+#include "lsst/log/Log.h"
+#include "proto/replication.pb.h"
+#include "replica/Messenger.h"
+#include "replica/Performance.h"
+#include "replica/ProtocolBuffer.h"
+
+
+// This macro to appear witin each block which requires thread safety
+#define LOCK_GUARD std::lock_guard<std::mutex> lock(_mtx)
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.ServiceManagementRequest");
+
+namespace proto = lsst::qserv::proto;
+
+/// Dump a collection of request descriptions onto the output stream
+void dumpRequestInfo(std::ostream& os,
+                     std::vector<proto::ReplicationServiceResponseInfo> const& requests) {
+
+    for (const auto &r : requests) {
+        os  << "\n"
+            << "    type:     " << proto::ReplicationReplicaRequestType_Name(r.replica_type()) << "\n"
+            << "    id:       " << r.id() << "\n"
+            << "    priority: " << r.priority() << "\n"
+            << "    database: " << r.database() << "\n";
+
+        switch (r.replica_type()) {
+            case proto::ReplicationReplicaRequestType::REPLICA_CREATE:
+                os  << "    chunk:    " << r.chunk() << "\n"
+                    << "    worker:   " << r.worker() << "\n";
+                break;
+
+            case proto::ReplicationReplicaRequestType::REPLICA_DELETE:
+            case proto::ReplicationReplicaRequestType::REPLICA_FIND:
+                os  << "    chunk:    " << r.chunk() << "\n";
+                break;
+
+            case proto::ReplicationReplicaRequestType::REPLICA_FIND_ALL:
+                break;
+
+            default:
+                throw std::logic_error (
+                            "unhandled request type " +
+                            proto::ReplicationReplicaRequestType_Name(r.replica_type()) +
+                            " in  ServiceManagementRequest::dumpRequestInfo");
+        }
+    }
+}
+
+} /// namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+///////////////////////////////////
+//         ServiceState          //
+///////////////////////////////////
+
+void ServiceState::set(proto::ReplicationServiceResponse const& message) {
+
```

Comment:
no lock in here? I'm confused about the locking in Request subclasses.

Reply:
The ServiceState struct is defined in ServiceManagementRequestBase.h and it's used as private data member of class ServiceManagementRequestBase. Any modifications to the objects of this (ServiceState) class are thread-safe (see details inside ServiceManagementRequestBase.cc). The only access to ServiceState is guarded via a check to the required FINISHED (always forward) state of the ServiceManagementRequestBase. So, th accessor is also safe.